boa/src/builtins/map/mod.rs

```diff
@@ -0,0 +1,325 @@
+#![allow(clippy::mutable_key_type)]
+
+use super::function::{make_builtin_fn, make_constructor_fn};
+use crate::{
+    builtins::{
+        object::{ObjectData, INSTANCE_PROTOTYPE, PROTOTYPE},
+        property::Property,
+        value::{ResultValue, Value},
+    },
+    exec::Interpreter,
+    BoaProfiler,
+};
+use gc::GcCell;
+use ordered_map::OrderedMap;
+
+pub mod ordered_map;
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug, Clone)]
+pub(crate) struct Map(OrderedMap<Value, Value>);
+
+impl Map {
+    pub(crate) const NAME: &'static str = "Map";
+
+    pub(crate) const LENGTH: usize = 1;
+
+    /// Helper function to get the map object.
+    /*fn get_map(this: &Value, ctx: &mut Interpreter) -> Result<OrderedMap<Value, Value>, Value> {
+        if let Value::Object(ref object) = this {
+            let object = object.borrow();
+            if let Some(map) = object.as_map() {
+                return Ok(map);
+            }
+        }
+        Err(ctx.construct_type_error("'this' is not a Map"))
+    }*/
+
+    /// Helper function to set the size property.
+    fn set_size(this: &Value, size: usize) {
+        let size = Property::new()
+            .value(Value::from(size))
+            .writable(false)
+            .configurable(false)
+            .enumerable(false);
+
+        this.set_property("size".to_string(), size);
+    }
+
+    /// `Map.prototype.set( key, value )`
+    ///
+    /// This method associates the value with the key. Returns the map object.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-map.prototype.set
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/set
+    pub(crate) fn set(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultValue {
+        let (key, value) = match args.len() {
+            0 => (Value::Undefined, Value::Undefined),
+            1 => (args[0].clone(), Value::Undefined),
+            _ => (args[0].clone(), args[1].clone()),
+        };
+
+        let size = if let Value::Object(ref object) = this {
+            let object = object.borrow();
+            if let Some(map) = object.as_map_ref() {
+                let mut map = map.borrow_mut();
+                map.insert(key, value);
+                map.len()
+            } else {
+                return Err(ctx.construct_type_error("'this' is not a Map"));
+            }
+        } else {
+            return Err(ctx.construct_type_error("'this' is not a Map"));
+        };
```

Comment:
```suggestion
        let size = if let Value::Object(ref object) = this {
            let object = object.borrow_mut();
            if let Some(ref mut map) = object.as_map_ref() {
                map.insert(key, value);
                map.len()
            } else {
                return Err(ctx.construct_type_error("'this' is not a Map"));
            }
        } else {
            return Err(ctx.construct_type_error("'this' is not a Map"));
        };
```
Wouldn't this work without the `GcCell`?

Reply:
Doesn't as_map_ref only return an immutable reference though? The GcCell allows us to use interior mutability to modify it without cloning the data and then replacing the objectdata, which is what it was doing previously

Reply:
What we could do is move the GcCell into the OrderedMap so that the modification methods work on an immutable reference.

Reply:
> Doesn't as_map_ref only return an immutable reference though?

Yes. it does currently, but we can change it to take a muttable self and return a mutable map.

> The GcCell allows us to use interior mutability to modify it without cloning the data and then replacing the objectdata, which is what it was doing previously

The way the `Value` is setup `Value::Object` contains a `GcObject` this is just a wrapper for `Gc<GcCell<Object>>` so we can mutate value with immutable reference, so it already has `interior mutability`, this is why we can set a field to `this` with an immutable reference, like `set_size()` method

Reply:
Ah yes, I didn't see that. I'll do that then